File: Feed/src/main/java/com/j/jface/feed/fragments/DebugToolsFragment.java
package com.j.jface.feed.fragments;

import android.app.Fragment;
import android.content.Intent;
import android.net.Uri;
import android.os.Handler;
import android.os.Message;
import android.support.annotation.NonNull;
import android.text.format.Time;
import android.util.Log;
import android.view.View;
import android.widget.CheckBox;
import android.widget.NumberPicker;
import android.widget.TextView;
import android.widget.TimePicker;

import com.google.firebase.iid.FirebaseInstanceId;
import com.j.jface.Const;
import com.j.jface.R;
import com.j.jface.action.InformUserAction;
import com.j.jface.feed.views.SnackbarRegistry;
import com.j.jface.lifecycle.WrappedFragment;
import com.j.jface.org.todo.TodoProvider;
import com.j.jface.wear.Wear;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;

import kotlin.Unit;

public class DebugToolsFragment extends WrappedFragment implements View.OnClickListener, NumberPicker.OnValueChangeListener, TimePicker.OnTimeChangedListener
{
  private static final int MSG_UPDATE_TIME = 1;
  private static final int GRACE_FOR_UPDATE = 3000;

  @NonNull private final Wear mWear;
  @NonNull private final Fragment mFragment;
  private long mOffset = 0;
  private boolean mTicking = false;
  @NonNull private final Time mTime1, mTime2;
  @NonNull private final NumberPicker mDaysOffsetUI;
  @NonNull private final NumberPicker mHoursUI;
  @NonNull private final NumberPicker mMinutesUI;
  @NonNull private final NumberPicker mSecondsUI;
  @NonNull private final TextView mOffsetLabel;
  @NonNull private final CheckBox[] mFenceUIs;

  private static class TabDebugToolsHandler extends Handler
  {
    private final DebugToolsFragment p;
    private TabDebugToolsHandler(final DebugToolsFragment p) { this.p = p; }
    @Override public void handleMessage(@NonNull Message message)
    {
      switch (message.what)
      {
        case MSG_UPDATE_TIME:
          p.tick();
          sendEmptyMessageDelayed(MSG_UPDATE_TIME, 1000);
      }
    }
  }
  private final Handler mHandler = new TabDebugToolsHandler(this);
  public DebugToolsFragment(@NonNull final Args a, @NonNull final Wear b)
  {
    super(a.inflater.inflate(R.layout.fragment_debug_tools, a.container, false));
    mWear = b;
    mFragment = a.fragment;
    mTime1 = new Time(); mTime2 = new Time();
    mView.findViewById(R.id.button_now).setOnClickListener(this);
    mDaysOffsetUI = mView.findViewById(R.id.daysOffsetUI);
    mDaysOffsetUI.setMinValue(0); mDaysOffsetUI.setMaxValue(14);
    mHoursUI = mView.findViewById(R.id.hoursUI);
    mHoursUI.setMinValue(0); mHoursUI.setMaxValue(23);
    mMinutesUI = mView.findViewById(R.id.minutesUI);
    mMinutesUI.setMinValue(0); mMinutesUI.setMaxValue(59);
    mSecondsUI = mView.findViewById(R.id.secondsUI);
    mSecondsUI.setMinValue(0); mSecondsUI.setMaxValue(59);
    mOffsetLabel = mView.findViewById(R.id.offsetLabel);
    tick();
    mHandler.sendEmptyMessageDelayed(MSG_UPDATE_TIME, 1000);

    mHoursUI.setOnValueChangedListener(this);
    mMinutesUI.setOnValueChangedListener(this);
    mSecondsUI.setOnValueChangedListener(this);
    mDaysOffsetUI.setOnValueChangedListener(this);

    mFenceUIs = new CheckBox[4];
    mFenceUIs[0] = mView.findViewById(R.id.fence1);
    mFenceUIs[1] = mView.findViewById(R.id.fence2);
    mFenceUIs[2] = mView.findViewById(R.id.fence3);
    mFenceUIs[3] = mView.findViewById(R.id.fence4);
    for (final CheckBox c : mFenceUIs) c.setOnClickListener(this);
    if (Const.RIO_MODE) mFenceUIs[2].setText("六本木");

    final TextView nodeNameTextView = mView.findViewById(R.id.nodeId_textView);
      mWear.getNodeName(mFragment.getContext(), nodeName -> {
        a.fragment.getActivity().runOnUiThread(() -> nodeNameTextView.setText("Node id : " + nodeName));
        return Unit.INSTANCE;
     });

    mView.findViewById(R.id.button_copy_todo_from_storage).setOnClickListener(v ->
    {
      final Intent intent = new Intent();
      intent.setType("*/*");
      intent.setAction(Intent.ACTION_GET_CONTENT);
      a.fragment.startActivityForResult(intent, Const.DESTROY_DATABASE_AND_REPLACE_WITH_FILE_CONTENTS_RESULT_CODE);
    });

    mView.findViewById(R.id.button_delete_FCM_token).setOnClickListener(v ->
    {
      new Thread(() ->
      {
        try { FirebaseInstanceId.getInstance().deleteInstanceId(); } catch (IOException e) { Log.e("Can't delete token", "" + e); }
        FirebaseInstanceId.getInstance().getToken(); // Force token generation
      }).start();
    });
  }

  @Override protected void onResume()
  {
    SnackbarRegistry.INSTANCE.setSnackbarParent(mView);
  }

  @Override protected void onPause()
  {
    SnackbarRegistry.INSTANCE.unsetSnackbarParent(mView);
  }

  protected void onActivityResult(final int requestCode, final int resultCode, final Intent data)
  {
    if (null == data || Const.DESTROY_DATABASE_AND_REPLACE_WITH_FILE_CONTENTS_RESULT_CODE != requestCode) return;
    final Uri fileUri = data.getData();
    FileNotFoundException exception = null;
    if (null != fileUri)
    {
      InputStream is = null;
      try { is = mFragment.getContext().getContentResolver().openInputStream(fileUri); }
      catch (final FileNotFoundException e) { exception = e; }
      if (null == exception && null != is)
        // Do this on the UI thread because YOLO. This is only for debug so we don't care.
        if (TodoProvider.destroyDatabaseAndReplaceWithFileContents(mFragment.getContext(), is))
        {
          new InformUserAction(mFragment.getActivity(), "Database dumped, restart JOrg", "Kill", v -> System.exit(0), null).invoke();
          return;
        }
    }
    new InformUserAction(mFragment.getActivity(), "File can't be opened." + (null == exception ? "" : " " + exception.toString()), null, null, null).invoke();
  }

  private void tick()
  {
    mTicking = true;
    mTime1.set(System.currentTimeMillis() + mOffset);
    mHoursUI.setValue(mTime1.hour);
    mMinutesUI.setValue(mTime1.minute);
    mSecondsUI.setValue(mTime1.second);

    mTime2.setToNow();
    mOffsetLabel.setText(Long.toString(mOffset));
    mTicking = false;
  }

  private void updateOffset(final int grace)
  {
    if (mTicking) return;
    mHandler.removeMessages(MSG_UPDATE_TIME);
    mTime2.setToNow();
    final int second = mSecondsUI.getValue();
    final int minute = mMinutesUI.getValue();
    final int hour = mHoursUI.getValue();
    mTime1.set(second, minute, hour, mTime2.monthDay, mTime2.month, mTime2.year);
    final int dayOffset = mDaysOffsetUI.getValue();
    mTime1.set(mTime1.toMillis(true) + dayOffset * 86400000 - grace);
    mOffset = mTime1.toMillis(true) - mTime2.toMillis(true);
    mHandler.sendEmptyMessageDelayed(MSG_UPDATE_TIME, grace);
    mWear.putDataToCloud(Const.DATA_PATH + "/" + Const.DATA_KEY_DEBUG_TIME_OFFSET, Const.DATA_KEY_DEBUG_TIME_OFFSET, mOffset);
  }

  private void updateFences()
  {
    int fences = 0;
    for (int i = 0; i < mFenceUIs.length; ++i)
      fences |= mFenceUIs[i].isChecked() ? 1 << i : 0;
    mWear.putDataToCloud(Const.DATA_PATH + "/" + Const.DATA_KEY_DEBUG_FENCES, Const.DATA_KEY_DEBUG_FENCES, fences);
  }

  @Override public void onClick(final View v)
  {
    switch (v.getId())
    {
      case R.id.button_now:
        mOffset = 0;
        tick();
        updateOffset(0);
        mHandler.removeMessages(MSG_UPDATE_TIME);
        mHandler.sendEmptyMessage(MSG_UPDATE_TIME);
      default:
        updateFences();
    }
  }

  @Override public void onValueChange(final NumberPicker picker, final int oldVal, final int newVal)
  {
    updateOffset(GRACE_FOR_UPDATE);
  }

  @Override public void onTimeChanged(final TimePicker view, final int hourOfDay, final int minute)
  {
    updateOffset(GRACE_FOR_UPDATE);
  }
}

File: Feed/src/main/java/com/j/jface/feed/fragments/DebugToolsFragment.kt
package com.j.jface.feed.fragments

import android.os.Handler
import android.os.Message
import android.text.format.DateUtils
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.CheckBox
import android.widget.NumberPicker
import android.widget.TextView
import android.widget.TimePicker
import com.google.android.gms.wearable.DataMap
import com.google.firebase.messaging.FirebaseMessaging
import com.j.jface.Const
import com.j.jface.R
import com.j.jface.feed.views.FaceView
import com.j.jface.feed.views.SnackbarRegistry
import com.j.jface.firebase.Firebase
import com.j.jface.lifecycle.WrappedFragment
import com.j.jface.wear.Wear
import kotlinx.coroutines.MainCoroutineDispatcher
import java.io.IOException
import java.time.Instant
import java.time.LocalTime
import java.time.OffsetDateTime
import java.time.ZoneId

const val MSG_UPDATE_TIME = 1
const val DEBUG_FENCES_PATH = "${Const.DATA_PATH}/${Const.DATA_KEY_DEBUG_FENCES}"
const val DEBUG_TIME_OFFSET_PATH = "${Const.DATA_PATH}/${Const.DATA_KEY_DEBUG_TIME_OFFSET}"

class DebugToolsFragment(a : WrappedFragment.Args, private val mWear : Wear) : WrappedFragment(a.inflater.inflate(R.layout.fragment_debug_tools, a.container, false)), View.OnClickListener, NumberPicker.OnValueChangeListener, TimePicker.OnTimeChangedListener
{
  private val mFragment = a.fragment
  private var mOffset = 0L
  private var mTicking = false
  private val mDaysOffsetUI : NumberPicker = mView.findViewById(R.id.daysOffsetUI)
  private val mHoursUI : NumberPicker = mView.findViewById(R.id.hoursUI)
  private val mMinutesUI : NumberPicker = mView.findViewById(R.id.minutesUI)
  private val mSecondsUI : NumberPicker = mView.findViewById(R.id.secondsUI)
  private val mOffsetLabel : TextView = mView.findViewById(R.id.offsetLabel)
  private val mFenceUIs = arrayOf(R.id.fence1, R.id.fence2, R.id.fence3, R.id.fence4).map { mView.findViewById<CheckBox>(it) }
  private val mHandler = TabDebugToolsHandler(this)
  private val mFace : FaceView = mView.findViewById(R.id.face_simulator) as FaceView
  private val mFaceToggles = arrayOf(R.id.debug_btn_coarse, R.id.debug_btn_burnin, R.id.debug_btn_ambient, R.id.debug_btn_visible).map { mView.findViewById<Button>(it) }
  private val mWearUpdateListener = object : Firebase.WearDataUpdateListener() {
    override fun onWearDataUpdated(path : String, data : DataMap) = this@DebugToolsFragment.onWearDataUpdated(path, data)
  }

  private class TabDebugToolsHandler(private val p : DebugToolsFragment) : Handler()
  {
    override fun handleMessage(message : Message)
    {
      when (message.what)
      {
        MSG_UPDATE_TIME ->
        {
          p.tick()
          sendEmptyMessageDelayed(MSG_UPDATE_TIME, 1000)
        }
      }
    }
  }

  init
  {
    mView.findViewById<View>(R.id.button_now).setOnClickListener(this)
    mDaysOffsetUI.minValue = 0
    mDaysOffsetUI.maxValue = 14
    mHoursUI.minValue = 0
    mHoursUI.maxValue = 23
    mMinutesUI.minValue = 0
    mMinutesUI.maxValue = 59
    mSecondsUI.minValue = 0
    mSecondsUI.maxValue = 59
    tick()
    mHandler.sendEmptyMessageDelayed(MSG_UPDATE_TIME, 1000)

    addListeners()

    if (Const.RIO_MODE) mFenceUIs[2].text = "渋谷"

    val nodeNameTextView = mView.findViewById<TextView>(R.id.nodeId_textView)
    mWear.getNodeName(mFragment.context!!) { nodeName ->
      a.fragment.activity!!.runOnUiThread { nodeNameTextView.text = "Node id : ${nodeName}" }
    }

    mView.findViewById<View>(R.id.button_delete_FCM_token).setOnClickListener {
      Thread {
        try
        {
          FirebaseMessaging.getInstance().deleteToken()
        }
        catch (e : IOException)
        {
          Log.e("Can't delete token", "" + e)
        }

        FirebaseMessaging.getInstance().token // Force token generation
      }.start()
    }
  }

  private fun addListeners()
  {
    mHoursUI.setOnValueChangedListener(this)
    mMinutesUI.setOnValueChangedListener(this)
    mSecondsUI.setOnValueChangedListener(this)
    mDaysOffsetUI.setOnValueChangedListener(this)
    mFenceUIs.forEach { it.setOnClickListener(this) }
    mFaceToggles.forEach { it.setOnClickListener(this) }
  }

  private fun removeListeners()
  {
    mHoursUI.setOnValueChangedListener(null)
    mMinutesUI.setOnValueChangedListener(null)
    mSecondsUI.setOnValueChangedListener(null)
    mDaysOffsetUI.setOnValueChangedListener(null)
    mFenceUIs.forEach { it.setOnClickListener(null) }
    mFaceToggles.forEach { it.setOnClickListener(null) }
  }

  private fun <R> withoutListeners(block : () -> R)
  {
    try
    {
      removeListeners()
      block()
    }
    finally
    {
      addListeners()
    }
  }

  override fun onResume()
  {
    SnackbarRegistry.setSnackbarParent(mView)
    mWearUpdateListener.resume()
  }

  override fun onPause()
  {
    SnackbarRegistry.unsetSnackbarParent(mView)
    mWearUpdateListener.pause()
  }

  override fun onDestroy() {
    mFace.onDestroy()
  }

  private fun tick()
  {
    mTicking = true
    val now = System.currentTimeMillis() + mOffset
    val zdt = OffsetDateTime.ofInstant(Instant.ofEpochMilli(now), ZoneId.systemDefault())
    val offset = mOffset
    mTicking = false

    mFragment.activity?.runOnUiThread {
      mHoursUI.value = zdt.hour
      mMinutesUI.value = zdt.minute
      mSecondsUI.value = zdt.second
      mOffsetLabel.text = java.lang.Long.toString(offset)
    }
  }

  private fun baseDay(timestamp : Long, utcOffset : Long) : Long = (timestamp + utcOffset * 1000) / DateUtils.DAY_IN_MILLIS

  fun onWearDataUpdated(path : String, data : DataMap)
  {
    when (path)
    {
      DEBUG_TIME_OFFSET_PATH -> withoutListeners {
        data.get<Long>(Const.DATA_KEY_DEBUG_TIME_OFFSET)?.let { mOffset = it }
        val now = System.currentTimeMillis()
        val utcOffset = Const.MILLISECONDS_TO_UTC
        mDaysOffsetUI.value = (baseDay(now + mOffset, utcOffset) - baseDay(now, utcOffset)).toInt()
        tick()
      }
      DEBUG_FENCES_PATH -> withoutListeners {
          val fences = data.get<Long>(Const.DATA_KEY_DEBUG_FENCES)?.toInt() ?: 0
          for (i in mFenceUIs.indices) mFenceUIs[i].isChecked = 0 != (fences and (1 shl i))
      }
    }
    mFace.invalidate();
  }

  private fun updateOffset()
  {
    if (mTicking) return
    mHandler.removeMessages(MSG_UPDATE_TIME)
    val now = System.currentTimeMillis()
    val offsetTime = OffsetDateTime.ofInstant(Instant.ofEpochMilli(now), ZoneId.systemDefault()).with(LocalTime.of(mHoursUI.value, mMinutesUI.value, mSecondsUI.value)).toEpochSecond() * 1000 + now % 1000
    val dayOffset = mDaysOffsetUI.value
    val offsetDateTime = offsetTime + dayOffset * 86400000
    mOffset = offsetDateTime - now
    mHandler.sendEmptyMessage(MSG_UPDATE_TIME)
    mWear.putDataToCloud(DEBUG_TIME_OFFSET_PATH, Const.DATA_KEY_DEBUG_TIME_OFFSET, mOffset)
  }

  private fun updateFences()
  {
    var fences = 0
    for (i in mFenceUIs.indices)
      if (mFenceUIs[i].isChecked)
        fences = fences or (1 shl i)
    mWear.putDataToCloud(DEBUG_FENCES_PATH, Const.DATA_KEY_DEBUG_FENCES, fences.toLong())
  }

  override fun onClick(v : View)
  {
    when (v.id)
    {
      R.id.button_now ->
      {
        mOffset = 0
        tick()
        updateOffset()
        mHandler.removeMessages(MSG_UPDATE_TIME)
        mHandler.sendEmptyMessage(MSG_UPDATE_TIME)
        updateFences()
      }
      R.id.debug_btn_coarse -> mFace.toggleCoarse()
      R.id.debug_btn_burnin -> mFace.toggleBurnin()
      R.id.debug_btn_ambient -> mFace.toggleAmbient()
      R.id.debug_btn_visible -> mFace.toggleVisible()
      else -> updateFences()
    }
    mFace.invalidate();
  }

  override fun onValueChange(picker : NumberPicker, oldVal : Int, newVal : Int) = updateOffset()
  override fun onTimeChanged(view : TimePicker, hourOfDay : Int, minute : Int) = updateOffset()
}
